Introduction of Provider Networks

johannwagner opened the pull request:
<!--
    Thank you for your interest in contributing to Nautobot! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->
### Fixes: #724
<!--
    Please include a summary of the proposed changes below.
-->

This pull request introduces the functionality of Provider Networks. They are known from Netbox.

Note: This is a WIP and it is not fully functional.

TODOs:

- [ ] Migrate Tests
- [ ] Confirm full functionality of Provider Networks

johannwagner commented:
Currently, I am blocked by https://github.com/nautobot/nautobot/issues/1137 since I am not able to test this feature properly, I guess.

johannwagner commented on nautobot/circuits/models.py:
```diff
@@ -9,16 +10,54 @@
+
+
+@extras_features("custom_fields", "custom_links", "export_templates", "webhooks")
+class ProviderNetwork(PrimaryModel):
```
Will implement this after the initial port of this feature!

chadell replied:
thanks!

glennmatthews commented:
Hi @johannwagner, it appears that #1137 was fixed by #1145. If you update your branch to pull in latest `develop` content, you should be able to proceed with testing. Thank you for taking on this effort!

johannwagner commented:
I rebased my work and it seems to work similar to the Netbox implementation.

I think, I want to keep this seperate. We want to implement @chadell idea for our own purposes but I would like to merge this PR first and add the IP Range association within another PR.

johannwagner commented:
What is needed to merge this?

chadell commented on nautobot/circuits/forms.py:
```diff
@@ -223,7 +279,7 @@ class CircuitFilterForm(BootstrapMixin, TenancyFilterForm, StatusFilterFormMixin
         "type",
         "provider",
-        "status",
+        "providernetwork" "status",
```
typo?

johannwagner replied:
Unsure, ProviderNetwork is not part of circuits, therefore I removed it.

chadell commented on nautobot/circuits/migrations/0005_auto_20211209_1421.py:
```diff
@@ -0,0 +1,30 @@
+import sys
```
could we merge migrations and are a meaningful name?

chadell commented on nautobot/circuits/models.py:
```diff
@@ -9,16 +10,54 @@
+
+@extras_features("custom_fields", "custom_links", "export_templates", "webhooks")
+class ProviderNetwork(PrimaryModel):
+    name = models.CharField(max_length=100)
```
maybe support `slug` field? and use it instead of `pk`?

johannwagner replied:
Would you mind elaborate this? We could add a slug field, but until now, only organizational objects have a slug, e.g. Regions, Providers, Sites, Device Types, and not specific "implementations" of organizational objects.

chadell commented on nautobot/circuits/models.py:
```diff
@@ -9,16 +10,54 @@
+    "ProviderNetwork",
 )
 
+from ..utilities.querysets import RestrictedQuerySet
```
```suggestion
from nautobot.utilities.querysets import RestrictedQuerySet
```

chadell commented on nautobot/dcim/tests/test_models.py:
```diff
@@ -539,6 +544,14 @@ def test_cable_cannot_terminate_to_an_existing_connection(self):
         with self.assertRaises(ValidationError):
             cable.clean()
 
+    def test_cable_cannot_terminate_to_a_cloud_circuittermination(self):
```
```suggestion
    def test_cable_cannot_terminate_to_a_providernetwork_circuittermination(self):
```

chadell commented on nautobot/dcim/tests/test_models.py:
```diff
@@ -539,6 +544,14 @@ def test_cable_cannot_terminate_to_an_existing_connection(self):
 
+    def test_cable_cannot_terminate_to_a_cloud_circuittermination(self):
+        """
+        Neither side of a cable can be terminated to a CircuitTermination which is attached to a Cloud
```
```suggestion
        Neither side of a cable can be terminated to a CircuitTermination which is attached to a ProviderNetwork
```

chadell commented on nautobot/circuits/templates/circuits/circuittermination_edit.html:
```diff
@@ -40,8 +40,21 @@ <h3>{% block title %}{{ obj.circuit.provider }} {{ obj.circuit }} - Side {{ form
+                                      {% render_field form.region %}
+                                      {% render_field form.site %}
+                                  </div>
+                                  <div class="tab-pane{% if providernetwork_tab_active %} active{% endif %}" id="cloud">
```
`s/cloud/providernetwork` ?

chadell commented on nautobot/circuits/templates/circuits/circuittermination_edit.html:
```diff
@@ -40,8 +40,21 @@ <h3>{% block title %}{{ obj.circuit.provider }} {{ obj.circuit }} - Side {{ form
+                                      {% render_field form.site %}
+                                  </div>
+                                  <div class="tab-pane{% if providernetwork_tab_active %} active{% endif %}" id="cloud">
+                                      {% render_field form.cloud %}
```
`s/cloud/providernetwork` ?

chadell commented on nautobot/circuits/tests/test_api.py:
```diff
@@ -46,6 +46,46 @@ def setUpTestData(cls):
+        )
+        Provider.objects.bulk_create(providers)
+
+        clouds = (
```
```suggestion
        provider_networks = (
```

chadell commented on nautobot/circuits/tests/test_api.py:
```diff
@@ -46,6 +46,46 @@ def setUpTestData(cls):
+            ProviderNetwork(name="Provider Network 2", provider=providers[0]),
+            ProviderNetwork(name="Provider Network 3", provider=providers[0]),
+        )
+        ProviderNetwork.objects.bulk_create(clouds)
```
```suggestion
        ProviderNetwork.objects.bulk_create(provider_networks)
```

chadell commented on nautobot/circuits/api/nested_serializers.py:
```diff
@@ -8,8 +8,21 @@
+    url = serializers.HyperlinkedIdentityField(view_name="circuits-api:providernetwork-detail")
+
+    class Meta:
+        model = Provider
```
souldn't this be `ProviderNetwork`?
```suggestion
        model = ProviderNetwork
```

chadell reviewed (changes requested):
Comments from a first review (not complete)

glennmatthews commented on nautobot/circuits/navigation.py:
```diff
@@ -82,6 +82,28 @@
+                    NavMenuItem(
+                        link="circuits:providernetwork_list",
+                        name="Provider Networks",
+                        weight=100,
```
```suggestion
                        weight=200,
```

glennmatthews commented on nautobot/circuits/templates/circuits/providernetwork.html:
```diff
@@ -0,0 +1,68 @@
+  <li>{{ object }}</li>
+{% endblock %}
+
+{% block content %}
```
It would be preferable to override the `content_left_page`, `content_right_page`, and/or `content_full_width_page` sub-blocks rather than the entire `content` block, as that reduces the amount of boilerplate required in this file.

johannwagner replied:
Implemetend!

glennmatthews commented on nautobot/circuits/tests/test_filters.py:
```diff
@@ -237,6 +248,11 @@ def test_provider(self):
         params = {"provider": [provider.slug]}
         self.assertEqual(self.filterset(params, self.queryset).qs.count(), 3)
 
+    def test_providernetwork(self):
+        providernetwork = ProviderNetwork.objects.all()[:2]
+        params = {"providernetwork_id": [providernetwork[0].pk, providernetwork[1].pk]}
```
```suggestion
    def test_provider_network(self):
        provider_network = ProviderNetwork.objects.all()[:2]
        params = {"provider_network_id": [provider_network[0].pk, provider_network[1].pk]}
```

glennmatthews commented on nautobot/circuits/tests/test_filters.py:
```diff
@@ -410,6 +444,11 @@ def test_circuit_id(self):
         params = {"circuit_id": [circuits[0].pk, circuits[1].pk]}
         self.assertEqual(self.filterset(params, self.queryset).qs.count(), 4)
 
+    def test_providernetwork(self):
+        providernetworks = ProviderNetwork.objects.all()[:2]
+        params = {"providernetwork_id": [providernetworks[0].pk, providernetworks[1].pk]}
```
```suggestion
    def test_provider_network(self):
        provider_networks = ProviderNetwork.objects.all()[:2]
        params = {"provider_network_id": [provider_networks[0].pk, provider_networks[1].pk]}
```

glennmatthews commented on nautobot/circuits/urls.py:
```diff
@@ -42,6 +42,22 @@
         name="provider_changelog",
         kwargs={"model": Provider},
     ),
+    path("providernetworks/", views.ProviderNetworkListView.as_view(), name="providernetwork_list"),
+    path("providernetworks/add/", views.ProviderNetworkEditView.as_view(), name="providernetwork_add"),
+    path("providernetworks/import/", views.ProviderNetworkBulkImportView.as_view(), name="providernetwork_import"),
+    path("providernetworks/edit/", views.ProviderNetworkBulkEditView.as_view(), name="providernetwork_bulk_edit"),
+    path("providernetworks/delete/", views.ProviderNetworkBulkDeleteView.as_view(), name="providernetwork_bulk_delete"),
+    path("providernetworks/<uuid:pk>/", views.ProviderNetworkView.as_view(), name="providernetwork"),
+    path("providernetworks/<uuid:pk>/edit/", views.ProviderNetworkEditView.as_view(), name="providernetwork_edit"),
+    path(
+        "providernetworks/<uuid:pk>/delete/", views.ProviderNetworkDeleteView.as_view(), name="providernetwork_delete"
+    ),
+    path(
+        "providernetworks/<uuid:pk>/changelog/",
```
```suggestion
    path("provider-networks/", views.ProviderNetworkListView.as_view(), name="providernetwork_list"),
    path("provider-networks/add/", views.ProviderNetworkEditView.as_view(), name="providernetwork_add"),
    path("provider-networks/import/", views.ProviderNetworkBulkImportView.as_view(), name="providernetwork_import"),
    path("provider-networks/edit/", views.ProviderNetworkBulkEditView.as_view(), name="providernetwork_bulk_edit"),
    path("provider-networks/delete/", views.ProviderNetworkBulkDeleteView.as_view(), name="providernetwork_bulk_delete"),
    path("provider-networks/<uuid:pk>/", views.ProviderNetworkView.as_view(), name="providernetwork"),
    path("provider-networks/<uuid:pk>/edit/", views.ProviderNetworkEditView.as_view(), name="providernetwork_edit"),
    path(
        "provider-networks/<uuid:pk>/delete/", views.ProviderNetworkDeleteView.as_view(), name="providernetwork_delete"
    ),
    path(
        "provider-networks/<uuid:pk>/changelog/",
```

glennmatthews commented on nautobot/circuits/views.py:
```diff
@@ -78,6 +78,68 @@ class ProviderBulkDeleteView(generic.BulkDeleteView):
+    def get_extra_context(self, request, instance):
+        circuits = (
+            Circuit.objects.restrict(request.user, "view")
+            .filter(Q(termination_a__providernetwork=instance.pk) | Q(termination_z__providernetwork=instance.pk))
```
```suggestion
            .filter(Q(termination_a__provider_network=instance.pk) | Q(termination_z__provider_network=instance.pk))
```

glennmatthews commented on nautobot/docs/models/circuits/circuittermination.md:
```diff
@@ -2,9 +2,9 @@
+Each circuit termination is attached to either a site or a provider network. Site terminations may optionally be connected via a cable to a specific device interface or port within that site. Each termination must be assigned a port speed, and can optionally be assigned an upstream speed if it differs from the downstream speed (a common scenario with e.g. DOCSIS cable modems). Fields are also ava ...
 
-In adherence with Nautobot's philosophy of closely modeling the real world, a circuit may terminate only to a physical interface. For example, circuits may not terminate to LAG interfaces, which are virtual in nature. In such cases, a separate physical circuit is associated with each LAG member interface and each needs to be modeled discretely.
+In adherence with Nautobot's philosophy of closely modeling the real world, a circuit be connected only to a physical interface. For example, circuits may not terminate to LAG interfaces, which are virtual in nature. In such cases, a separate physical circuit is associated with each LAG member interface and each needs to be modeled discretely.
```
```suggestion
In adherence with Nautobot's philosophy of closely modeling the real world, a circuit may be connected only to a physical interface. For example, circuits may not terminate to LAG interfaces, which are virtual in nature. In such cases, a separate physical circuit is associated with each LAG member interface and each needs to be modeled discretely.
```

glennmatthews reviewed (commented):
Thank you for the extensive pull request!

johannwagner commented on nautobot/circuits/tables.py:
```diff
@@ -9,7 +9,24 @@
+
+    class Meta(BaseTable.Meta):
+        model = ProviderNetwork
+        fields = ("pk", "name", "provider", "description", "tags")
```
I am not sure about this, because comments can contain Markdown. I think, this should not be shown within a table. `description` can be shown and this is only text.

johannwagner replied:
Also, it is not done anywhere else.

glennmatthews replied:
Good call, thanks!

glennmatthews commented on nautobot/circuits/api/serializers.py:
```diff
@@ -25,6 +25,7 @@
     NestedCircuitTerminationSerializer,
     NestedCircuitTypeSerializer,
     NestedProviderSerializer,
+    NestedProviderNetworkSerializer,
```
Alphabetization nitpick

```suggestion
    NestedProviderNetworkSerializer,
    NestedProviderSerializer,
```

glennmatthews commented on nautobot/circuits/api/serializers.py:
```diff
@@ -59,6 +60,34 @@ class Meta:
+        fields = [
+            "id",
+            "url",
+            "display",
```
"display" is not generally listed explicitly as a field here.

```suggestion
```

glennmatthews commented on nautobot/circuits/api/serializers.py:
```diff
@@ -59,6 +60,34 @@ class Meta:
+            "url",
+            "display",
+            "provider",
+            "name",
+            "description",
```
```suggestion
            "name",
            "slug",
            "description",
```

glennmatthews commented on nautobot/circuits/api/nested_serializers.py:
```diff
@@ -1,15 +1,28 @@
+
+    class Meta:
+        model = ProviderNetwork
+        fields = ["id", "url", "display", "name"]
```
```suggestion
        fields = ["id", "url", "name", "slug"]
```

glennmatthews commented on nautobot/circuits/filters.py:
```diff
@@ -72,6 +73,35 @@ def search(self, queryset, name, value):
+
+    class Meta:
+        model = ProviderNetwork
+        fields = ["id", "name"]
```
```suggestion
        fields = ["id", "name", "slug"]
```

glennmatthews commented on nautobot/circuits/models.py:
```diff
@@ -9,17 +10,64 @@
+    "webhooks",
+)
+class ProviderNetwork(PrimaryModel):
+    name = models.CharField(max_length=100, unique=True)
```
I think `unique=True` is incorrect here as below you're defining the `unique_together = ("provider", "name")` - it should be permissible to use the same name across different providers, I think?

johannwagner replied:
Since we are using this as a base for our slug, which needs to be unique, we may also want to enforce it here. 
If it is no problem for the slug handling, I am happy to remove it.

glennmatthews replied:
`AutoSlugField` is supposed to automatically handle uniqueness at creation time - IIRC if you create Provider 1 Network "Alpha" (slug `alpha`) and Provider 2 Network "Alpha" (slug `alpha`) the latter will automatically become slug `alpha1` so as to have a unique slug - but by all means please try it out to confirm.

johannwagner replied:
It does not get automagically changed, but the creation gives a reasonable error.

glennmatthews commented on nautobot/circuits/models.py:
```diff
@@ -9,17 +10,64 @@
+            self.description,
+            self.comments,
+        )
+
```
Suggest adding a custom `display` implementation for the REST API:

```suggestion

    @property
    def display(self):
        return f"{self.provider} {self.name}"
```

glennmatthews commented on nautobot/circuits/models.py:
```diff
@@ -223,7 +273,20 @@ def termination_z(self):
+    )
+    provider_network = models.ForeignKey(
+        to="circuits.ProviderNetwork",
+        on_delete=models.PROTECT,
```
@jathanism @lampwins Thoughts on using `PROTECT` vs `CASCADE` here?

glennmatthews commented:
Getting nearly there!

Shoot - I just realized that this should be a PR against `next` rather than `develop` since it's a new feature. Can you please make any necessary changes to your branch and retarget the PR accordingly?

Thank you!

johannwagner commented:
Okay, should there be any needed changes? I switched the base branch and it seems to merge fine. I did not test it, will do tomorrow.

johannwagner commented:
I will also fix tests tomorrow.

johannwagner commented:
Unit tests work on my local machine, hopefully CI is also happy about them.

glennmatthews commented:
Looks like the migrations file needs to be renumbered/regenerated so as not to conflict with the migration introduced in #1107. Sorry for the trouble!

jathanism commented on nautobot/circuits/migrations/0004_providernetworks.py:
```diff
@@ -0,0 +1,142 @@
+            name="providernetwork",
+            unique_together={("provider", "name")},
+        ),
+        migrations.RunPython(
+            code=cache_circuit_terminations,
+            reverse_code=django.db.migrations.operations.special.RunPython.noop,
+        ),
+        migrations.AddField(
+            model_name="providernetwork",
+            name="slug",
+            field=nautobot.core.fields.AutoSlugField(blank=True, max_length=100, populate_from="name", unique=True),
+        ),
+    ]
```
It's not immediately clear to me why the `AddField` for `provider network.slug` is after the `RunPython` data migration. Was this intentional or can we get away with leaving this in this migration file and moving the `RunPython` to its own migration file that immediately follows these schema changes?

I'm asking because in order to split this it would have to be either:

1. Add `ProviderNetwork` in `0004_providernetwork`
2. Run `cache_circuit_terminations` data migration in `0005_cache_circuit_terminations`
3. Run `AddField` to add `ProviderNetwork.slug` in `0006_providernetwork_slug`

Or ideally, just the first two:

1. Add `ProviderNetwork` (including `slug`) in `0004_providernetwork`
2. Run `cache_circuit_terminations` data migration in `0005_cache_circuit_terminations`

johannwagner replied:
This was not intentional, I fixed it as you mentioned.

jathanism commented on nautobot/circuits/models.py:
```diff
@@ -240,7 +307,21 @@ class Meta:
-        return "Side {}".format(self.get_term_side_display())
+        return f"Termination {self.term_side}: {self.site or self.provider_network}"
+
+    def get_absolute_url(self):
+        if self.site:
+            return self.site.get_absolute_url()
+        return self.provider_network.get_absolute_url()
```
I'm not clear on why we'd want the `absolute_url` to be for the `site` if it's set. I see in other code context how it's conditional, however, can you please add a comment of the intent here for posterity?

jathanism commented on nautobot/circuits/tables.py:
```diff
@@ -9,7 +9,31 @@
+from .models import Circuit, CircuitType, Provider, ProviderNetwork
+
+CIRCUIT_TERMINATION_PARENT = """
+{% if value.provider_network %}
+    <a href="{{ value.provider_network.get_absolute_url }}">{{ value.provider_network }}</a>
+{% elif value.site %}
+    <a href="{{ value.site.get_absolute_url }}">{{ value.site }}</a>
+{% endif %}
```
Seeing as the return value of `ProviderNetwork.get_absolute_url()` is already based on whether `ProviderNetwork.site` is set, this could be streamlined.
```suggestion
<a href="{{ value.provider_network.get_absolute_url }}">
{% if value.provider_network %}
    {{ value.provider_network }}
{% elif value.site %}
    {{ value.site }}
{% endif %}
</a>
```

jathanism reviewed (changes requested):
Really great contribution but I have some specific requests. Thanks!

johannwagner commented:
@jathanism Thanks for your comments. I mostly only ported the code from Netbox, so I am not that deep in all thought processes but I will do my best to resolve your concerns.

johannwagner commented:
Okay, I added the feedback from @jathanism and rebased the branch to next for a better overview. I think, I need another CI run to make sure, everything is sorted now.

jathanism reviewed (approved):
Thank you for taking the time to port this over from the NetBox community. Great work!

johannwagner commented:
I enjoyed this one! Thanks for the great community experience and your time and effort! Looking forward to ship this!

glennmatthews reviewed (approved):
Great work!
